### PR TITLE
Async improvements

### PR DESCRIPTION
**Pull request by dom96**
From changelog:

- `newAsyncSocket` taking an `AsyncFD` now runs `setBlocking(false)` on the
  fd.

- The `ReadyKey` type in the selectors module now contains an ``errorCode``
  field to help distinguish between ``Event.Error`` events.
- <del>The `AsyncFD` type now reflects the fact that the underlying FD is registered
  in the async dispatcher.</del>
- Implemented an `accept` proc that works on a `SocketHandle` in
  ``nativesockets``.
- Implemented ``getIoHandler`` proc in the ``asyncdispatch`` module that allows
  you to retrieve the underlying IO Completion Port or ``Selector[AsyncData]``
  object in the specified dispatcher.

### CODE REVIEW

File: lib/pure/asyncdispatch.nim [hunk 1]
```diff
@@ -1091,10 +1108,17 @@ else:
       setGlobalDispatcher(newDispatcher())
     result = gDisp
 
-  proc register*(fd: AsyncFD) =
+  proc getIoHandler*(disp: PDispatcher): Selector[AsyncData] =
+    return disp.selector
+
+  proc register*(fd: cint | SocketHandle | AsyncFD): AsyncFD {.discardable.} =
```
Comment (Araq) [inline lib/pure/asyncdispatch.nim]: Questionable API design: It either returns AsyncFD(0) or ``fd`` which I already know. Return a discardable bool instead?

File: lib/pure/asyncfile.nim [hunk 2]
```diff
@@ -81,11 +81,10 @@ proc getFileSize*(f: AsyncFile): int64 =
   else:
     result = lseek(f.fd.cint, 0, SEEK_END)
 
-proc newAsyncFile*(fd: AsyncFd): AsyncFile =
+proc newAsyncFile*(fd: cint | AsyncFd): AsyncFile =
   ## Creates `AsyncFile` with a previously opened file descriptor `fd`.
   new result
-  result.fd = fd
-  register(result.fd)
+  result.fd = register(fd)
```
Comment (Araq) [inline lib/pure/asyncfile.nim]: Confusing chaining.

File: lib/pure/asyncnet.nim [hunk 3]
```diff
@@ -140,9 +140,18 @@ proc newAsyncSocket*(fd: AsyncFD, domain: Domain = AF_INET,
     sockType: SockType = SOCK_STREAM,
     protocol: Protocol = IPPROTO_TCP, buffered = true): AsyncSocket =
   ## Creates a new ``AsyncSocket`` based on the supplied params.
+  ##
+  ## The supplied ``fd``'s non-blocking state will be enabled implicitly.
+  ##
+  ## **Note**: This procedure will **NOT** register ``fd`` with the global
```
Comment (Araq) [inline lib/pure/asyncnet.nim]: This note is good but it indicates an API design problem.

File: lib/pure/ioselects/ioselectors_epoll.nim [hunk 4]
```diff
@@ -141,7 +141,7 @@ template checkFd(s, f) =
   if f >= s.maxFD:
     raiseIOSelectorsError("Maximum number of descriptors is exhausted!")
 
-proc registerHandle*[T](s: Selector[T], fd: SocketHandle,
+proc registerHandle*[T](s: Selector[T], fd: int | SocketHandle,
```
Comment (Araq) [inline lib/pure/ioselects/ioselectors_epoll.nim]: The point of the ``SocketHandle`` abstraction is type safety. ``int | SocketHandle`` is **not** typesafe.

File: lib/pure/ioselects/ioselectors_select.nim [hunk 5]
```diff
@@ -451,3 +451,6 @@ template withData*[T](s: Selector[T], fd: SocketHandle|int, value,
     else:
       body2
 
+
+proc getFd*[T](s: Selector[T]): int =
```
Comment (Araq) [inline lib/pure/ioselects/ioselectors_select.nim]: Shouldn't this be a ``SocketHandle``?

**Comment by Araq** [comment]
All these  ``int | SocketHandle`` weaken the type checking too much. Remove them. If the types are properly designed, there are not many conversions to do.

File: lib/pure/asyncdispatch.nim (hunk 1 above)
Comment (dom96) [reply lib/pure/asyncdispatch.nim]: I decided to give the type ``AsyncFD`` a meaning: it is an FD that is registered in the async dispatcher. This is why this function returns an ``AsyncFD``.

File: lib/pure/asyncfile.nim (hunk 2 above)
Comment (dom96) [reply lib/pure/asyncfile.nim]: You can think of this as: ``register`` takes an FD and returns a "registered" instance of that FD (cint -> AsyncFD in this case).

File: lib/pure/ioselects/ioselectors_select.nim (hunk 5 above)
Comment (dom96) [reply lib/pure/ioselects/ioselectors_select.nim]: A selector isn't a Socket. Again, I go back to the fact that this type is a misnomer. Perhaps I should rename it?

File: lib/pure/asyncfile.nim (hunk 2 above)
Comment (Araq) [reply lib/pure/asyncfile.nim]: Yes, I do undestand the code. But I dislike its design.

File: lib/pure/ioselects/ioselectors_select.nim (hunk 5 above)
Comment (Araq) [reply lib/pure/ioselects/ioselectors_select.nim]: Rename it then, yes.

File: lib/pure/asyncdispatch.nim (hunk 1 above)
Comment (Araq) [reply lib/pure/asyncdispatch.nim]: Why can I register an AsyncFD then? Am I allowed to register things multiple times?

Comment (dom96) [reply lib/pure/asyncdispatch.nim]: For backward compatibility. [If you pass in an AsyncFD then you won't get an error if its already in the dispatcher](https://github.com/nim-lang/Nim/pull/6796/files#diff-e7d29941cb2ca6bf4e0f5c627239bc7bR277)

File: lib/pure/asyncnet.nim (hunk 3 above)
Comment (dom96) [reply lib/pure/asyncnet.nim]: I will keep most of this note and create a separate PR to address this API design problem.

**Comment by dom96** [comment]
I removed the ``cint | AsyncFd | SocketHandle``, I will create a separate PR to sort out these types.

> All these int | SocketHandle weaken the type checking too much. Remove them. If the types are properly designed, there are not many conversions to do.

I decided to keep this for now. Many procedures already take ``int | SocketHandle`` so this makes the rest consistent. In a separate PR I will make a concrete type for file descriptors to be used instead of this generic. Hopefully you'll be okay with merging this as it stands.

**Comment by Araq** [comment]
Acceptable if the tests are green.

**Pull request merged**